### PR TITLE
Implement SMTP email functionality

### PR DESCRIPTION
**Pull request by zhongnansu**
### Description
- Implement SMTP email functionality
- Made minor fixes to webhook functionality and make send message to webhook IT working.

### Issues Resolved
#213 

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass, including unit test, integration test and doctest
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
  - [ ] New functionality has user manual doc added
- [ ] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: notifications/spi/src/main/kotlin/org/opensearch/notifications/spi/client/DestinationEmailClient.kt [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+ *
+ *  The OpenSearch Contributors require contributions made to
+ *  this file be licensed under the Apache-2.0 license or a
+ *  compatible open source license.
+ *
+ *  Modifications Copyright OpenSearch Contributors. See
+ *  GitHub history for details.
+ */
+
+/*
```
Comment (davidcui-amzn) [inline notifications/spi/src/main/kotlin/org/opensearch/notifications/spi/client/DestinationEmailClient.kt:12]: I think we don't need this second copyright statement for new files that we create

Comment (zhongnansu) [reply notifications/spi/src/main/kotlin/org/opensearch/notifications/spi/client/DestinationEmailClient.kt:12]: OK, new files? Do you mean new files after we migrated to OpenSearch?

Comment (davidcui-amzn) [reply notifications/spi/src/main/kotlin/org/opensearch/notifications/spi/client/DestinationEmailClient.kt:12]: Yeah

Comment (zhongnansu) [reply notifications/spi/src/main/kotlin/org/opensearch/notifications/spi/client/DestinationEmailClient.kt]: Removed unwanted headers in new file of this PR. Will use another PR later to fix all header in this project

**Review by davidcui-amzn — APPROVED** [review]

File: notifications/spi/src/main/kotlin/org/opensearch/notifications/spi/client/DestinationEmailClient.kt [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+        val prop = Properties()
+        prop["mail.transport.protocol"] = "smtp"
+        prop["mail.smtp.host"] = emailDestination.host
+        prop["mail.smtp.port"] = emailDestination.port
+        val session = Session.getInstance(prop)
+
+        when (emailDestination.method) {
+            "ssl" -> prop["mail.smtp.ssl.enable"] = true
+            "starttls" -> prop["mail.smtp.starttls.enable"] = true
```
Comment (joshuali925) [inline notifications/spi/src/main/kotlin/org/opensearch/notifications/spi/client/DestinationEmailClient.kt:59]: what's the naming logic for `start_tls` vs `starttls`? Also I believe no encryption sometimes appears as `none` and other times `plain`?

File: notifications/spi/src/main/kotlin/org/opensearch/notifications/spi/client/DestinationEmailClient.kt [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+            SecurityAccess.doPrivileged { sendMessage(mimeMessage) }
+            log.info("Email-SMTP sent")
+            DestinationMessageResponse(RestStatus.OK, "Success")
+        } catch (exception: SendFailedException) {
+            DestinationMessageResponse(RestStatus.BAD_GATEWAY, getMessagingExceptionText(exception))
+        } catch (exception: MailConnectException) {
+            DestinationMessageResponse(RestStatus.SERVICE_UNAVAILABLE, getMessagingExceptionText(exception))
+        } catch (exception: MessagingException) {
+            DestinationMessageResponse(RestStatus.FAILED_DEPENDENCY, getMessagingExceptionText(exception))
+        }
```
Comment (joshuali925) [inline notifications/spi/src/main/kotlin/org/opensearch/notifications/spi/client/DestinationEmailClient.kt:85]: is there a more generic exception handler for other exceptions somewhere in the call stack?

File: notifications/spi/src/main/kotlin/org/opensearch/notifications/spi/client/DestinationHttpClient.kt [hunk 4]
```diff
@@ -125,20 +127,18 @@ class DestinationHttpClient {
...
     private fun getHttpResponse(destination: WebhookDestination, message: MessageContent): CloseableHttpResponse {
-        val httpRequest = HttpPost()
-        val uri = destination.buildUri()
+        val httpRequest = HttpPost(destination.url)
 
         if (destination is CustomWebhookDestination) {
             if (destination.headerParams.isEmpty()) {
                 // set default header
-                httpRequest.setHeader("Content-Type", "application/json")
+                httpRequest.setHeader("Content-type", "application/json")
```
Comment (joshuali925) [inline notifications/spi/src/main/kotlin/org/opensearch/notifications/spi/client/DestinationHttpClient.kt:135]: why did this change? I believe content type header is `Content-Type`

File: notifications/spi/src/main/kotlin/org/opensearch/notifications/spi/client/DestinationHttpClient.kt [hunk 5]
```diff
@@ -163,14 +163,18 @@ class DestinationHttpClient {
...
         if (!VALID_RESPONSE_STATUS.contains(statusCode)) {
             throw IOException("Failed: $response")
         }
     }
 
-    private fun extractBody(message: MessageContent): String {
-        return message.textDescription
+    private fun buildRequestBody(destination: WebhookDestination, message: MessageContent): String {
+        return if (destination is SlackDestination) {
+            "{\"text\" : \"${message.buildWebhookMessage()}\"}"
```
Comment (joshuali925) [inline notifications/spi/src/main/kotlin/org/opensearch/notifications/spi/client/DestinationHttpClient.kt:175]: any reference why slack is different?

File: notifications/spi/src/main/kotlin/org/opensearch/notifications/spi/model/destination/EmailDestination.kt [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+/**
+ * This class holds the contents of email destination
+ */
+class EmailDestination(
+    val host: String,
+    val port: Int,
+    val method: String,
+    val fromAddress: String,
+    val recipient: String,
+    destinationType: String,
```
Comment (joshuali925) [inline notifications/spi/src/main/kotlin/org/opensearch/notifications/spi/model/destination/EmailDestination.kt:42]: what's destinationType?

**Review by joshuali925 — APPROVED** [review]
a few questions

File: notifications/spi/src/main/kotlin/org/opensearch/notifications/spi/client/DestinationHttpClient.kt (hunk 5 above)
Comment (zhongnansu) [reply notifications/spi/src/main/kotlin/org/opensearch/notifications/spi/client/DestinationHttpClient.kt:175]: https://api.slack.com/messaging/webhooks
`text` is a requried field name

File: notifications/spi/src/main/kotlin/org/opensearch/notifications/spi/client/DestinationHttpClient.kt (hunk 4 above)
Comment (zhongnansu) [reply notifications/spi/src/main/kotlin/org/opensearch/notifications/spi/client/DestinationHttpClient.kt:135]: header Field names are case-insensitive, but I might have modified it by mistake, will keep consistancy

File: notifications/spi/src/main/kotlin/org/opensearch/notifications/spi/client/DestinationEmailClient.kt (hunk 2 above)
Comment (zhongnansu) [reply notifications/spi/src/main/kotlin/org/opensearch/notifications/spi/client/DestinationEmailClient.kt:59]: good catch, will change to start_tls to be consistant with data model, and add none as a condition

File: notifications/spi/src/main/kotlin/org/opensearch/notifications/spi/model/destination/EmailDestination.kt (hunk 6 above)
Comment (zhongnansu) [reply notifications/spi/src/main/kotlin/org/opensearch/notifications/spi/model/destination/EmailDestination.kt:42]: I'll add comments

Comment (joshuali925) [reply notifications/spi/src/main/kotlin/org/opensearch/notifications/spi/model/destination/EmailDestination.kt:42]: When should `smpt` and `smtp_account` be used? in contrast to `Smtp`
https://github.com/opensearch-project/notifications/blob/cdf2156c2eda8b44edab5880afe5ac8350454524/notifications/src/main/kotlin/org/opensearch/notifications/settings/EmailChannelType.kt#L33-L36

https://github.com/opensearch-project/common-utils/blob/HEAD/src/main/kotlin/org/opensearch/commons/notifications/model/ConfigType.kt#L60-L64
```kotlin
    SMTP_ACCOUNT("smtp_account") {
        override fun toString(): String {
            return tag
        }
    },
```

Comment (zhongnansu) [reply notifications/spi/src/main/kotlin/org/opensearch/notifications/spi/model/destination/EmailDestination.kt:42]: I think when creating `config type`, it's a `email` type first with `smtp_account` as childtype. But here in SPI we have different data model for destination, and we use `destination type` to distinguish. If sending smtp email, sender will import data class from spi and init an `EmailDestination` object(with destination type as member var) to pass to the SPI interface to send message.

Comment (joshuali925) [reply notifications/spi/src/main/kotlin/org/opensearch/notifications/spi/model/destination/EmailDestination.kt:42]: I see, thanks

**Pull request merged**